tools/wake/timeline.cpp

```diff
@@ -0,0 +1,242 @@
+    long dependency = -1;
+    for (auto access: accesses) {
+        dependency = fill_one_dependency(access, dependency, job_map);
+    }
+}
+
+void assign_parents(std::map<long, JobNode> &job_map, std::vector<long> &top_sorted_jobs) {
+    for (long job_id: top_sorted_jobs) {
+        JobNode &job_node = job_map.at(job_id);
+        for (long dependency_id: job_node.dependencies) {
+            JobNode dependency = job_map.at(dependency_id);
+            if (dependency.hypothetical_endtime >= job_node.hypothetical_starttime) {
+                job_node.hypothetical_starttime = dependency.hypothetical_endtime;
+                job_node.parent_id = dependency_id;
+            }
+        }
+
+        job_node.hypothetical_endtime = job_node.hypothetical_starttime + (job_node.endtime - job_node.starttime);
+    }
+}
+
+void find_critical_path(std::map<long, JobNode> &job_map, std::vector<long> &critical_path) {
+    long latest_job_id = -1;
+    for (const auto &job_pair: job_map) {
+        if (latest_job_id == -1 ||
+            job_map.at(latest_job_id).hypothetical_endtime < job_pair.second.hypothetical_endtime) {
+            latest_job_id = job_pair.second.job.job;
+        }
+    }
+    job_map.at(latest_job_id).is_on_critical_path = true;
+    critical_path.emplace_back(job_map.at(latest_job_id).job.job);
+    while (job_map.at(latest_job_id).parent_id != -1) {
+        latest_job_id = job_map.at(latest_job_id).parent_id;
+        job_map.at(latest_job_id).is_on_critical_path = true;
+        critical_path.emplace_back(job_map.at(latest_job_id).job.job);
+    }
+}
+
+void dfs_transitive_reduction(long job_id, std::map<long, JobNode> &job_map, std::vector<int> &ancestors) {
+    JobNode &job = job_map.at(job_id);
+    job_map.at(job_id).top_sort_visited = true;
+    ancestors.emplace_back(job_id);
+
+    for (long child_id: job.dependencies) {
+        JobNode &child = job_map.at(child_id);
+        for (long grandchild_id: child.dependencies) {
+            for (int ancestor_id: ancestors) {
+                JobNode ancestor = job_map.at(ancestor_id);
+                if (std::find(ancestor.dependencies.begin(), ancestor.dependencies.end(), grandchild_id) !=
+                    ancestor.dependencies.end()) {
+                    std::remove(ancestor.dependencies.begin(), ancestor.dependencies.end(), grandchild_id);
+                }
+            }
+        }
+        dfs_transitive_reduction(child_id, job_map, ancestors);
+    }
+    ancestors.pop_back();
+}
+
+void transitive_reduction(std::map<long, JobNode> &job_map) {
+    std::vector<int> ancestors;
+    for (const auto &job_pair: job_map) {
+        if (!(job_pair.second.transitive_reduction_visited)) {
+            dfs_transitive_reduction(job_pair.first, job_map, ancestors);
+        }
+    }
+}
+
+void write_jobs(std::map<long, JobNode> &job_map, std::ostream &os) {
```

Comment:
All of these write_foo functions can take their maps/vectors as const references

Reply:
Done